src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1096 @@
+	}
+
+	pid_t pid = { 0 };
+	WaitForBackgroundWorkerStartup(handle, &pid);
+
+	if (segOut)
+	{
+		*segOut = seg;
+	}
+
+	return handle;
+}
+
+
+typedef struct CitusBackgroundJobExecuterErrorCallbackContext
+{
+	const char *database;
+	const char *username;
+} CitusBackgroundJobExecuterErrorCallbackContext;
+
+
+static void
+CitusBackgroundJobExecuterErrorCallback(void *arg)
+{
+	CitusBackgroundJobExecuterErrorCallbackContext *context =
+		(CitusBackgroundJobExecuterErrorCallbackContext *) arg;
+	errcontext("Citus Background Job Executor: %s/%s", context->database,
+			   context->username);
+}
+
+
+/*
+ * Background worker logic.
+ *
+ * based on the background worker logic in pgcron
+ */
+void
+CitusBackgroundJobExecuter(Datum main_arg)
+{
+	/*
+	 * TODO figure out if we need this signal handler that is in pgcron
+	 * pqsignal(SIGTERM, pg_cron_background_worker_sigterm);
+	 */
+	BackgroundWorkerUnblockSignals();
+
+	/* Set up a memory context and resource owner. */
+	Assert(CurrentResourceOwner == NULL);
+	CurrentResourceOwner = ResourceOwnerCreate(NULL, "citus background job");
+	CurrentMemoryContext = AllocSetContextCreate(TopMemoryContext,
+												 "citus background job execution",
+												 ALLOCSET_DEFAULT_MINSIZE,
+												 ALLOCSET_DEFAULT_INITSIZE,
+												 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* Set up a dynamic shared memory segment. */
+	dsm_segment *seg = dsm_attach(DatumGetInt32(main_arg));
+	if (seg == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("unable to map dynamic shared memory segment")));
+	}
+
+	shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC, dsm_segment_address(seg));
+	if (toc == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("bad magic number in dynamic shared memory segment")));
+	}
+
+	char *database = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_DATABASE, false);
+	char *username = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_USERNAME, false);
+	char *command = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_COMMAND, false);
+	shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+
+	shm_mq_set_sender(mq, MyProc);
+	shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+	pq_redirect_to_shm_mq(seg, responseq);
+
+	/* setup error context to indicate the errors came from a running background task */
+	ErrorContextCallback errorCallback = { 0 };
+	CitusBackgroundJobExecuterErrorCallbackContext context = {
+		.database = database,
+		.username = username,
+	};
+	errorCallback.callback = CitusBackgroundJobExecuterErrorCallback;
+	errorCallback.arg = (void *) &context;
+	errorCallback.previous = error_context_stack;
+	error_context_stack = &errorCallback;
+
+	BackgroundWorkerInitializeConnection(database, username, 0);
+
+	/* Prepare to execute the query. */
+	SetCurrentStatementStartTimestamp();
+	debug_query_string = command;
+	pgstat_report_activity(STATE_RUNNING, command);
```

Comment:
it would be good to also set appname, because these processes are currently not easily discernible in pg_stat_activity